### ISSUE TITLE
assign @current_user not available in eex template

### ISSUE THREAD

**Issue by hakunin** [smpallen99/coherence#203]
Hey guys,
sorry for dumb question, its probably something mis-configured in my app, but somehow I don't have the router redirecting me when user is logged out and get an error instead.

```
  pipeline :protected do
    plug :accepts, ["html"]
    plug :fetch_session
    plug :fetch_flash
    plug :protect_from_forgery
    plug :put_secure_browser_headers
    plug Coherence.Authentication.Session, login: true # tried protected: true as well
  end

  scope "/", BlockViz.Web do
    pipe_through :protected

    resources "/wallets", WalletController
  end
```

any ideas?

**Comment by smpallen99** [comment]
Please review the Getting started section of the readme. You need both a :browser and :protected pipeline as well as two scopes calling `coherence_routes`, one for each :browser and :protected.

**Comment by hakunin** [comment]
I have that, this is the whole file:

```
defmodule BlockViz.Web.Router do
  use BlockViz.Web, :router
  use Coherence.Router

  pipeline :browser do
    plug :accepts, ["html"]
    plug :fetch_session
    plug :fetch_flash
    plug :protect_from_forgery
    plug :put_secure_browser_headers
    plug Coherence.Authentication.Session
  end

  pipeline :protected do
    plug :accepts, ["html"]
    plug :fetch_session
    plug :fetch_flash
    plug :protect_from_forgery
    plug :put_secure_browser_headers
    plug Coherence.Authentication.Session, login: true
  end

  pipeline :api do
    plug :accepts, ["json"]
  end

  # Coherence routes -----------------------

  scope "/" do
    pipe_through :browser
    coherence_routes()
  end

  scope "/" do
    pipe_through :protected
    coherence_routes :protected
  end

  # Our routes -----------------------------

  # Public resources
  scope "/", BlockViz.Web do
    pipe_through :browser

    get "/", PageController, :index
  end

  # Protected resources
  scope "/", BlockViz.Web do
    pipe_through :protected

    resources "/wallets", WalletController

    get "/connect", PageController, :connect
    get "/auth/coinbase", PageController, :callback
    post "/auth/coinbase", PageController, :callback

    scope "/auth" do
      get "/:provider", AuthController, :request
      get "/:provider/callback", AuthController, :callback
      post "/identity/callback", AuthController, :identity_callback
    end

    resources "/users", UserController
  end

end
```

I also tried including `plug Coherence.RequireLogin` in that controller but it didn't require the login either and still fails inside the template.

**Comment by hakunin** [comment]
I re-added the sessioncontroller without login to the protected pipeline and now it works:

```
  pipeline :protected do
    plug :accepts, ["html"]
    plug :fetch_session
    plug :fetch_flash
    plug :protect_from_forgery
    plug :put_secure_browser_headers
    plug Coherence.Authentication.Session
    plug Coherence.Authentication.Session, login: true
  end
```

In getting started I don't see that being done:
https://www.dropbox.com/s/cva2ycf2qesujk2/Screenshot%202017-05-13%2021.34.57.png?dl=0

But it works.

**Comment by smpallen99** [comment]
That suggests you have another issue in your router. Can you post your complete router.ex file please. I'll review it and give you some feedback on the implications on how you have it configured.

**Comment by smpallen99** [comment]
You should not have to  add the 2nd plug as you did. The documentation is correct. So something else is going on here. It would help if you posted the error you received. 

What version of Coherence? Have you updated coherence from an earlier version. I suspect that your missing the Redirect module module in your application. It is generated from the [template here](https://github.com/smpallen99/coherence/blob/master/priv/templates/coherence.install/controllers/coherence/redirects.ex). If you don't have the `web/controllers/coherence/redirects.ex` file in your project, you can just copy that template into that location. If that works, let me know. If it doesn't work, then let me know what version of coherence you are using and post the error.

Thanks...Steve

**Comment by hakunin** [comment]
I added coherence from branch `phx-1.3` to a fresh project. I have the redirects module, just like it was generated and have not modified it.

This is the error: https://www.dropbox.com/s/1rfl1ybvr0l3tao/Screenshot%202017-05-13%2022.12.03.png?dl=0

The double inclusion of session fixing it is really weird..

**Comment by smpallen99** [comment]
Ok, that what I needed. The problem is that current_user does not get set in the conn.assigns if you are not logged in. That is why I created `Coherence.current_user(@conn)`. So use that API, or you you don't want to, replace `@current_user` with `@conn.assigns[:current_user]`

**Comment by hakunin** [comment]
Thanks, still curious why double-including the session worked? I am trying to debug it by opening it up from dependencies, but `IO.inspect` doesn't print anything. Any tips? I might run into other issues I might want to debug so would love to know how to debug this kind of case.

**Comment by hakunin** [comment]
Looks like I found the problem.

This code will either remember where to return and call login, or assign user data, if there is no login callback.

```
  defp assert_login({conn, nil}, login, _) when is_function(login) do
    user_return_to =
      case conn.query_string do
        "" -> conn.request_path
        _ -> conn.request_path <> "?" <> conn.query_string
      end

    conn
    |> put_session("user_return_to",  user_return_to)
    |> login.()
  end
  defp assert_login({conn, user_data}, _, assign_key) do
    conn
    |> assign_user_data(user_data, assign_key)
    |> create_user_token(user_data, Config.user_token, assign_key)
  end
```

**Comment by smpallen99** [comment]
Not really a problem. Thats the way I designed it. I'll think about your suggestion on setting @current_user to nil, but there is something nagging me about this change. Not sure what yet. 

I designed this to use the `Coherence.current_user/1` API. It also uses your configured :current_user setting, so if you changes this, your layout would still work. 

If you want to modify something in deps, you just need to do `mix deps.compile coherence` to get it  to pick up your changes. If your doing a lot of this, then the article you lined would be a good approach.
